pkg/server/delta/v3/server.go

```diff
@@ -44,6 +132,340 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
+			}
+			nonce, err := send(resp, resource.RouteType)
+			if err != nil {
+				return err
+			}
+			values.deltaRouteNonce = nonce
+			setState(resp, resource.RouteType)
+		case resp, more := <-values.deltaListeners:
+			if !more {
+				return status.Errorf(codes.Unavailable, "listeners watch failed")
+			}
+			nonce, err := send(resp, resource.ListenerType)
+			if err != nil {
+				return err
+			}
+			values.deltaListenerNonce = nonce
+			setState(resp, resource.ListenerType)
+		case resp, more := <-values.deltaSecrets:
+			if !more {
+				return status.Errorf(codes.Unavailable, "secrets watch failed")
+			}
+			nonce, err := send(resp, resource.SecretType)
+			if err != nil {
+				return err
+			}
+			values.deltaSecretNonce = nonce
+			setState(resp, resource.SecretType)
+		case resp, more := <-values.deltaRuntimes:
+			if !more {
+				return status.Errorf(codes.Unavailable, "runtimes watch failed")
+			}
+			nonce, err := send(resp, resource.RuntimeType)
+			if err != nil {
+				return err
+			}
+			values.deltaRuntimeNonce = nonce
+			setState(resp, resource.RuntimeType)
+		case resp, more := <-values.deltaResponses:
+			if more {
+				if resp == deltaErrorResponse {
+					return status.Errorf(codes.Unavailable, "delta resource watch failed")
+				}
+				typeURL := resp.GetDeltaRequest().TypeUrl
+				nonce, err := send(resp, typeURL)
+				if err != nil {
+					return err
+				}
+				values.deltaNonces[typeURL] = nonce
+				setState(resp, typeURL)
+			}
+		case req, more := <-reqCh:
+			// input stream ended or errored out
+			if !more {
+				return nil
+			}
+			if req == nil {
+				return status.Errorf(codes.Unavailable, "empty request")
+			}
+
+			// Log out our error detail from  envoyif we get one but don't do anything crazy here yet
+			// TODO: embed a logger in the server so we can be a bit more verbose when needed
+			if req.ErrorDetail != nil {
+				return status.Errorf(codes.Code(req.ErrorDetail.GetCode()), "received error from xDS client: %s", req.ErrorDetail.GetMessage())
+			}
+
+			// node field in discovery request is delta-compressed
+			// nonces can be reused across streams; we verify nonce only if nonce is not initialized
+			var nonce string
+			if req.Node != nil {
+				node = req.Node
+				nonce = req.GetResponseNonce()
+			} else {
+				req.Node = node
+				// If we have no nonce, i.e. this is the first request on a delta stream, set one
+				nonce = strconv.FormatInt(streamNonce, 10)
+			}
+
+			// type URL is required for ADS but is implicit for xDS
+			if defaultTypeURL == resource.AnyType {
+				if req.TypeUrl == "" {
+					return status.Errorf(codes.InvalidArgument, "type URL is required for ADS")
+				}
+			} else if req.TypeUrl == "" {
+				req.TypeUrl = defaultTypeURL
+			}
+
+			state := values.deltaStreamStates[req.GetTypeUrl()]
+			// If this is empty we can assume this is the first time state
+			// is being set on this stream for this resource type
```

Comment:
Maybe something like `// Initialize the state map if we haven't already. This means that we're handling the first request for this type on the stream.`?

Reply:
I like that! Thank you